Intermittent integration test failure: test_check_targeting

Seen in https://app.circleci.com/pipelines/github/mozilla/experimenter/11357/workflows/ac5ffccc-a322-4747-b1d4-4598aab6859a/jobs/34418


```py
=================================== FAILURES ===================================
_______________ test_check_targeting[DESKTOP-first_run_win1903] ________________
[gw1] linux -- Python 3.8.10 /code/app/tests/integration/.tox/integration-test-nimbus/bin/python3

selenium = <selenium.webdriver.firefox.webdriver.WebDriver (session="c487ea42-9d28-44c2-9f48-0dccd469eaa2")>
base_url = 'https://nginx/nimbus/'
default_data = BaseExperimentDataClass(public_name='test_check_targeting[DESKTOPfirst_run_win1903]-ecc5', hypothesis='smart stuff her...e'>, min_version=80, targeting='first_run_win1903', percentage=50.0, expected_clients=50, locale=None, countries=None))
create_experiment = <function create_experiment.<locals>._create_experiment at 0x7f3d08675ee0>
app_data = 'first_run_win1903'
json_url = <function json_url.<locals>._json_url at 0x7f3d08523d30>

    @pytest.mark.run_parallel
    def test_check_targeting(
        selenium, base_url, default_data, create_experiment, app_data, json_url
    ):
        default_data.audience.targeting = app_data
        default_data.public_name = default_data.public_name.replace("-", "", 1)
        experiment = create_experiment(selenium)
        experiment.launch_to_preview()
>       experiment.wait_for_preview_status()

nimbus/test_remote_settings.py:143: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
nimbus/pages/experimenter/summary.py:61: in wait_for_preview_status
    self.wait_with_refresh(
nimbus/pages/base.py:27: in wait_with_refresh
    self.wait.until(_wait_for_it, message=message)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <selenium.webdriver.support.wait.WebDriverWait (session="c487ea42-9d28-44c2-9f48-0dccd469eaa2")>
method = <function Base.wait_with_refresh.<locals>._wait_for_it at 0x7f3d08523ca0>
message = 'Summary Page: Unable to find preview status'

    def until(self, method, message=''):
        """Calls the method provided with the driver as an argument until the \
        return value is not False."""
        screen = None
        stacktrace = None
    
        end_time = time.time() + self._timeout
        while True:
            try:
                value = method(self._driver)
                if value:
                    return value
            except self._ignored_exceptions as exc:
                screen = getattr(exc, 'screen', None)
                stacktrace = getattr(exc, 'stacktrace', None)
            time.sleep(self._poll)
            if time.time() > end_time:
                break
>       raise TimeoutException(message, screen, stacktrace)
E       selenium.common.exceptions.TimeoutException: Message: Summary Page: Unable to find preview status

.tox/integration-test-nimbus/lib/python3.8/site-packages/selenium/webdriver/support/wait.py:80: TimeoutException
------------------------------- pytest-selenium --------------------------------
Driver log: /tmp/pytest-of-seluser/pytest-0/popen-gw1/test_check_targeting_DESKTOP_f1/driver.log
URL: https://nginx/nimbus/test_check_targetingdesktopfirst_run_win1903-ecc5/
WARNING: Failed to gather log types: Message: HTTP method not allowed
=========================== short test summary info ============================
FAILED nimbus/test_remote_settings.py::test_check_targeting[DESKTOP-first_run_win1903]
=================== 1 failed, 15 passed in 92.30s (0:01:32) ====================
ERROR: InvocationError for command /code/app/tests/integration/.tox/integration-test-nimbus/bin/pytest --verify-base-url --base-url https://nginx/nimbus/ --self-contained-html --driver Firefox nimbus/ -k DESKTOP -m run_parallel -n 2 -vvv (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   integration-test-nimbus: commands failed
ERROR: 1
```

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1821)

--- a/app/tests/integration/nimbus/pages/experimenter/summary.py
+++ b/app/tests/integration/nimbus/pages/experimenter/summary.py
@@ -9,6 +9,7 @@ class SummaryPage(ExperimenterBase):
     """Experiment Summary Page."""
 
     _page_wait_locator = (By.CSS_SELECTOR, "#PageSummary")
+    _header_slug = (By.CSS_SELECTOR, 'p[data-testid="header-experiment-slug"]')
     _approve_request_button_locator = (By.CSS_SELECTOR, "#approve-request-button")
     _launch_to_preview_locator = (By.CSS_SELECTOR, "#launch-to-preview-button")
     _launch_without_preview_locator = (By.CSS_SELECTOR, "#launch-to-review-button")
@@ -100,6 +101,10 @@ class SummaryPage(ExperimenterBase):
             EC.presence_of_all_elements_located(self._clone_parent_locator),
             message="Summary Page: could not find clone parent",
         )
+
+    @property
+    def experiment_slug(self):
+        return self.wait_for_and_find_element(self._header_slug, "header slug").text
 
     @property
     def experiment_status(self):

--- a/app/tests/integration/nimbus/parallel_pytest_args.txt
+++ b/app/tests/integration/nimbus/parallel_pytest_args.txt
@@ -4,5 +4,4 @@
 -k KLAR -m run_per_app
 -k FOCUS -m run_per_app
 -k DESKTOP -m run_once
--k DESKTOP -m run_parallel -n 2
- 
+-k DESKTOP -m run_targeting -n 2

--- a/app/tests/integration/nimbus/test_firefox_targeting.py
+++ b/app/tests/integration/nimbus/test_firefox_targeting.py
@@ -1,0 +1,108 @@
+import json
+import time
+
+import pytest
+import requests
+from nimbus.models.base_dataclass import BaseExperimentApplications
+from nimbus.pages.browser import Browser
+
+LOAD_DATA_RETRIES = 10
+LOAD_DATA_RETRY_DELAY = 1.0
+
+
+def load_graphql_data(query):
+    for retry in range(LOAD_DATA_RETRIES):
+        try:
+            return requests.post(
+                "https://nginx/api/v5/graphql",
+                json=query,
+                verify=False,
+            ).json()
+        except json.JSONDecodeError:
+            if retry + 1 >= LOAD_DATA_RETRIES:
+                raise
+            else:
+                time.sleep(LOAD_DATA_RETRY_DELAY)
+
+
+def load_targeting_configs():
+    data = load_graphql_data(
+        {
+            "operationName": "getConfig",
+            "variables": {},
+            "query": """
+                    query getConfig {
+                        nimbusConfig {
+                            targetingConfigs {
+                                label
+                                value
+                                applicationValues
+                            }
+                        }
+                    }
+                    """,
+        }
+    )
+    targeting_configs = []
+    for item in data["data"]["nimbusConfig"]["targetingConfigs"]:
+        if "DESKTOP" in item["applicationValues"]:
+            targeting_configs.append(item["value"])
+    return targeting_configs
+
+
+def load_experiment_data(slug):
+    return load_graphql_data(
+        {
+            "operationName": "getExperiment",
+            "variables": {"slug": slug},
+            "query": """
+                    query getExperiment($slug: String!) {
+                        experimentBySlug(slug: $slug) {
+                            jexlTargetingExpression
+                            recipeJson
+                            __typename
+                        }
+                    }
+                    """,
+        }
+    )
+
+
+@pytest.fixture(params=load_targeting_configs())
+def targeting_config_slug(request):
+    return request.param
+
+
+@pytest.mark.run_targeting
+def test_check_targeting(
+    selenium,
+    default_data,
+    create_experiment,
+    targeting_config_slug,
+):
+    # TODO #6791
+    # If the targeting config slug includes the word desktop it will cause this test
+    # to run against applications other than desktop, which will then fail.
+    # This check will prevent the test from executing fully but we should dig
+    # into preventing this case altogether when we have time.
+    if default_data.application != BaseExperimentApplications.DESKTOP:
+        return
+
+    default_data.audience.targeting = targeting_config_slug
+    experiment = create_experiment(selenium)
+
+    experiment_data = load_experiment_data(experiment.experiment_slug)
+    targeting = experiment_data["data"]["experimentBySlug"]["jexlTargetingExpression"]
+    recipe = experiment_data["data"]["experimentBySlug"]["recipeJson"]
+
+    # Inject filter expression
+    selenium.get("about:blank")
+    with open("nimbus/utils/filter_expression.js") as js:
+        result = Browser.execute_script(
+            selenium,
+            targeting,
+            json.dumps({"experiment": recipe}),
+            script=js.read(),
+            context="chrome",
+        )
+    assert result is not None, "Invalid Targeting, or bad recipe"

--- a/app/tests/integration/nimbus/test_remote_settings.py
+++ b/app/tests/integration/nimbus/test_remote_settings.py
@@ -1,48 +1,7 @@
-import json
-import time
-
 import pytest
-import requests
-from nimbus.models.base_dataclass import BaseExperimentApplications
-from nimbus.pages.browser import Browser
 from nimbus.pages.experimenter.summary import SummaryPage
 from nimbus.pages.remote_settings.dashboard import Dashboard
 from nimbus.pages.remote_settings.login import Login
-
-LOAD_DATA_RETRIES = 10
-LOAD_DATA_RETRY_DELAY = 1.0
-
-
-def load_data():
-    for retry in range(0, LOAD_DATA_RETRIES):
-        try:
-            apps = []
-            data = requests.post(
-                "https://nginx/api/v5/graphql",
-                json={
-                    "operationName": "getConfig",
-                    "variables": {},
-                    "query": "\nquery getConfig {\n  nimbusConfig "
-                    "{\n    targetingConfigs {\n      "
-                    "label\n      value\n      applicationValues\n    "
-                    "}\n  }\n}\n",
-                },
-                verify=False,
-            ).json()
-            for item in data["data"]["nimbusConfig"]["targetingConfigs"]:
-                if "DESKTOP" in item["applicationValues"]:
-                    apps.append(item["value"])
-            return apps
-        except json.JSONDecodeError:
-            if retry + 1 >= LOAD_DATA_RETRIES:
-                raise
-            else:
-                time.sleep(LOAD_DATA_RETRY_DELAY)
-
-
-@pytest.fixture(params=load_data())
-def app_data(request):
-    return request.param
 
 
 @pytest.mark.run_per_app
@@ -131,39 +90,3 @@ def test_end_experiment_and_reject_end(
     Dashboard(selenium, kinto_review_url).open().reject()
 
     SummaryPage(selenium, experiment_url).open().wait_for_rejected_alert()
-
-
-@pytest.mark.run_parallel
-def test_check_targeting(
-    selenium, base_url, default_data, create_experiment, app_data, json_url
-):
-    # TODO #6791
-    # If the targeting config slug includes the word desktop it will cause this test
-    # to run against applications other than desktop, which will then fail.
-    # This check will prevent the test from executing fully but we should dig
-    # into preventing this case altogether when we have time.
-    if default_data.application != BaseExperimentApplications.DESKTOP:
-        return
-
-    default_data.audience.targeting = app_data
-    default_data.public_name = default_data.public_name.replace("-", "", 1)
-    experiment = create_experiment(selenium)
-    experiment.launch_to_preview()
-    experiment.wait_for_preview_status()
-    json_url = json_url(base_url, default_data.public_name)
-    # Get experiment JSON and parse
-    experiment_json = requests.get(f"{json_url}", verify=False).json()
-    experiment_json = {"experiment": experiment_json}
-    targeting = experiment_json["experiment"]["targeting"]
-    experiment_json = json.dumps(experiment_json)
-    # Inject filter expression
-    selenium.get("about:blank")
-    with open("nimbus/utils/filter_expression.js") as js:
-        script = Browser.execute_script(
-            selenium,
-            targeting,
-            experiment_json,
-            script=js.read(),
-            context="chrome",
-        )
-    assert script is not None, "Invalid Targeting, or bad recipe"

--- a/app/tests/integration/nimbus/utils/filter_expression.js
+++ b/app/tests/integration/nimbus/utils/filter_expression.js
@@ -1,8 +1,3 @@
-Components.utils.import("resource://gre/modules/components-utils/FilterExpressions.jsm");
-Components.utils.import("resource://gre/modules/components-utils/ClientEnvironment.jsm");
-
-// Targeting
-
 async function remoteSettings(arguments) {
 
     /*

--- a/app/tests/integration/tox.ini
+++ b/app/tests/integration/tox.ini
@@ -21,4 +21,4 @@ markers =
     use_variables: marks tests that need to use pytest-variables
     run_once: only run a single instance of this test (rather than for each app)
     run_per_app: run this test for each application
-    run_parallel: run these tests in parallel
+    run_targeting: run jexl targeting tests in firefox

